### PR TITLE
Minor tweaks to `Differentiable` protocol.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Conform `TangentVector` and `CotangentVector` to
  `VectorNumeric where Scalar : FloatingPoint`.
- Rename `moved(toward:)` to `moved(along:)`.

Also remove extraneous shape-related `VectorNumeric` requirements from
`FloatingPoint` types. Those requirements are no longer necessary now
that `ShapedVectorNumeric` has been split from `VectorNumeric`.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Review by rxwei — APPROVED** [review]
Nice!

File: stdlib/public/core/AutoDiff.swift [hunk 1]
```diff
@@ -60,23 +60,25 @@ public extension VectorNumeric {
 public protocol Differentiable {
   /// The tangent vector space of this differentiable manifold.
   associatedtype TangentVector : Differentiable
-    where TangentVector.TangentVector == TangentVector
+    where TangentVector.TangentVector == TangentVector,
+          TangentVector.Scalar : FloatingPoint
```
Comment (rxwei) [inline stdlib/public/core/AutoDiff.swift]: Regarding the build error: I think you need to specify `TangentVector`'s conformance to `VectorNumeric` here.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by rxwei** [comment]
Ah right, tests should be updated

**Comment by dan-zheng** [comment]
Updating tests soon!

Since the `TangentVector` and `CotangentVector` associated types must now conform to `VectorNumeric`, `Differentiable` types where `TangentVector = CotangentVector = Self` effectively must conform to `VectorNumeric` too.

I suppose this is fine and intentional. `Tensor where Scalar : Numeric` and `FloatingPoint` all conform to `VectorNumeric`.

File: test/AutoDiff/differentiable_attr_type_checking.swift [hunk 2]
```diff
@@ -540,10 +540,24 @@ struct JVPStruct {
...
+  static func * (lhs: Float, rhs: JVPStruct) -> JVPStruct {
+    return JVPStruct(p: lhs * rhs.p)
+  }
+}
+
 extension JVPStruct : Differentiable {
   typealias TangentVector = JVPStruct
   typealias CotangentVector = JVPStruct
-  func moved(toward direction: JVPStruct) -> JVPStruct {
+  func moved(along direction: JVPStruct) -> JVPStruct {
```
Comment (rxwei) [inline test/AutoDiff/differentiable_attr_type_checking.swift]: You can remove this and the next method. They provided by default when Tangent == Self

File: test/AutoDiff/protocol_requirement_autodiff.swift [hunk 3]
```diff
@@ -26,7 +26,7 @@ extension DiffReq {
 struct Quadratic : DiffReq, Equatable {
   typealias TangentVector = Quadratic
   typealias CoangentVector = Quadratic
-  func moved(toward q: Quadratic) -> Quadratic {
+  func moved(along q: Quadratic) -> Quadratic {
```
Comment (rxwei) [inline test/AutoDiff/protocol_requirement_autodiff.swift]: Same

File: stdlib/public/core/FloatingPointTypes.swift.gyb [hunk 4]
```diff
@@ -1832,12 +1832,7 @@ extension ${Self} : Strideable {
 //===----------------------------------------------------------------------===//
 
 extension ${Self} : VectorNumeric {
-  public typealias Shape = ()
```
Comment (rxwei) [inline stdlib/public/core/FloatingPointTypes.swift.gyb]: Awesome

File: test/AutoDiff/differentiable_attr_type_checking.swift (hunk 2 above)
Comment (dan-zheng) [reply test/AutoDiff/differentiable_attr_type_checking.swift]: Thanks! Done

File: test/AutoDiff/protocol_requirement_autodiff.swift (hunk 3 above)
Comment (dan-zheng) [reply test/AutoDiff/protocol_requirement_autodiff.swift]: Done

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**